nemo/core/classes/common.py

```diff
@@ -952,24 +1157,208 @@ def _get_hf_hub_pretrained_model_info(cls, model_name: str, refresh_cache: bool
-        is_token_available = HfFolder.get_token() is not None
-
-        # Try to load the model from the Huggingface Hub
-        path = hf_hub_download(
-            repo_id=model_name,
-            filename=resolved_model_filename,
-            library_name="nemo",
-            library_version=nemo.__version__,
-            force_download=refresh_cache,
-            use_auth_token=is_token_available,
-        )
+        hf_token = get_hf_token()
+
+        # First check if .nemo file exists in HF
+        api = HfApi(token=hf_token)
+
+        # Check if model exists in HF
+        nemo_file_exists = api.file_exists(repo_id=model_name, filename=resolved_model_filename, repo_type="model")
+
+        if nemo_file_exists:
+            # Try to load the model from the Huggingface Hub
+            path = hf_hub_download(
+                repo_id=model_name,
+                filename=resolved_model_filename,
+                library_name='nemo',
+                library_version=nemo.__version__,
+                force_download=refresh_cache,
+                token=hf_token,
+            )
+        else:
+            repo_info = api.repo_info(repo_id=model_name, token=hf_token)
+
+            # Download whole HF repo and load entire directory as nemo directory
+            cache_dir = Path.joinpath(resolve_cache_dir(), "hf_hub_cache", f'{model_name}')
+            # If either description and location in the cloud changes, this will force re-download
+            cache_subfolder = repo_info.sha
+            # if file exists on cache_folder/subfolder, it will be re-used, unless refresh_cache is True
+            save_path = os.path.join(cache_dir, cache_subfolder)
+
+            if os.path.exists(cache_dir):
+                num_files_in_dir = len(os.listdir(cache_dir))
+                print("Found {} files in cache directory {}".format(num_files_in_dir, cache_dir))
+                print(f"Deleting old cache directory for model `{model_name}` in order to prevent duplicates...")
+                shutil.rmtree(cache_dir, ignore_errors=True)
+
+            if not os.path.exists(save_path):
+                logging.info(f"Downloading {model_name} from HuggingFace Hub to path: {save_path}")
+                os.makedirs(save_path, exist_ok=True)
+
+            path = snapshot_download(
+                repo_id=model_name,
+                library_name='nemo',
+                library_version=nemo.__version__,
+                force_download=refresh_cache,
+                cache_dir=save_path,
+                local_dir=save_path,
+                local_dir_use_symlinks=False,
+                token=hf_token,
+            )
 
         # Cannot pre-resolve the specific class without double instantiation (first for config, second for model params)
         # Default to current class, and perform basic class path resolution (handled via restore_from() + target class)
         class_ = cls
 
         return class_, path
 
+    def push_to_hf_hub(
```

Comment:
Maybe all the hf functions should go into their own file?

Reply:
Hmm they are tied to FileIO - how should we refactor them ? Put them in other file and then import and wrap them inside of fileio?

Reply:
Done